### PR TITLE
Config validation and build target alignment

### LINKED ISSUE #170: Config Validation and Build Target Alignment
# Config Validation and Build Target Alignment

> Add upper-bound validation to numeric config values (`testTimeout`, `planTimeout`, `concurrency`) and fix the tsup build target to match the minimum Node.js runtime declared in `package.json`.

## Context

Dispatch is a TypeScript CLI tool (`src/cli.ts`) that orchestrates AI agents. It uses tsup for bundling, Vitest for testing, and stores persistent user configuration in `.dispatch/config.json`.

Configuration is managed in `src/config.ts`, which defines:
- `DispatchConfig` interface — the shape of persisted config values
- `CONFIG_KEYS` — the list of valid config key names
- `validateConfigValue()` — validates a string value for a given config key, returning `null` on success or an error message string on failure

Currently, `CONFIG_KEYS` only includes `["provider", "model", "source", "testTimeout"]`. The fields `planTimeout` and `concurrency` are parsed as CLI-only arguments in `src/cli.ts` (via `--plan-timeout` and `--concurrency` flags) but are **not** part of the config system — they cannot be set via `config.json` or validated by `validateConfigValue()`.

The CLI argument parser in `src/cli.ts` already enforces some bounds for `--concurrency` (must be a positive integer ≤ `MAX_CONCURRENCY` which is 64) and `--plan-timeout` (must be a positive number), but these are ad-hoc checks with no shared bounds constant.

The build config `tsup.config.ts` sets `target: "node18"` while `package.json` declares `engines.node: ">=20.12.0"`. The `tsconfig.json` targets `ES2022`, which is appropriate for Node 20.

Tests live in `src/tests/` and use Vitest. The existing config tests are in `src/tests/config.test.ts`. CLI argument parsing tests are in `src/tests/cli.test.ts`.

Key files involved:
- `src/config.ts` — config interface, keys, validation logic
- `src/cli.ts` — CLI argument parsing, `MAX_CONCURRENCY` constant
- `tsup.config.ts` — build configuration
- `tsconfig.json` — TypeScript compiler options
- `package.json` — engine requirements, scripts
- `src/tests/config.test.ts` — existing config unit tests
- `src/tests/cli.test.ts` — existing CLI tests (imports `MAX_CONCURRENCY`)
- `src/orchestrator/runner.ts` — `RawCliArgs` interface that includes `concurrency` and `planTimeout`

## Why

- **Unbounded numeric config values are dangerous.** `testTimeout` currently accepts any positive finite number. A user entering `1000` (intending seconds but getting minutes) would wait ~16 hours. `planTimeout` has the same risk. Upper bounds create a safety net and produce error messages that communicate the valid range.
- **Unbounded concurrency can crash the host.** Setting `concurrency = 200` would spawn hundreds of concurrent processes, exhausting file descriptors and memory. While the CLI has a `MAX_CONCURRENCY = 64` cap, this is not shared with config validation, and the config system doesn't even support `concurrency` as a key yet.
- **Build target mismatch.** `tsup.config.ts` targets `node18` but the project requires Node ≥ 20.12.0. This means tsup may down-level ES2022+ features that Node 20 supports natively, producing unnecessary polyfills or — worse — silently breaking if a polyfill is incomplete.
- **Centralizing bounds** in a single `CONFIG_BOUNDS` constant eliminates magic numbers scattered across CLI parsing and config validation, making limits discoverable and easy to change.

### 1. Extend the config system with `planTimeout` and `concurrency`

Add `planTimeout` (number, optional) and `concurrency` (number, optional) to the `DispatchConfig` interface in `src/config.ts`. Add both keys to the `CONFIG_KEYS` array. Add corresponding validation branches in `validateConfigValue()`.

### 2. Define a shared `CONFIG_BOUNDS` constant

Create a `CONFIG_BOUNDS` constant object in `src/config.ts` that defines `{ min, max }` for each bounded numeric field: `testTimeout` (1–120), `planTimeout` (1–120), `concurrency` (1–32). Export it so that both `validateConfigValue()` and CLI parsing can reference the same bounds. This replaces the hardcoded `MAX_CONCURRENCY = 64` in `src/cli.ts` with the centralized value.

### 3. Update validation logic

Modify each numeric validation branch in `validateConfigValue()` to check `value >= bounds.min && value <= bounds.max` and produce an error message that states the valid range. The existing positive-number check for `testTimeout` becomes a range check.

### 4. Update CLI parsing to use shared bounds

Refactor `src/cli.ts` to import `CONFIG_BOUNDS` from `src/config.ts` and use it for `--concurrency` and `--plan-timeout` validation instead of local constants/hardcoded values. Remove or alias `MAX_CONCURRENCY` to maintain backward compatibility with tests that import it.

### 5. Fix tsup build target

Change `target: "node18"` to `target: "node20"` in `tsup.config.ts`. Verify `tsconfig.json` is consistent (it already targets ES2022, which is fine for Node 20).

### 6. Add comprehensive boundary tests

Add tests to `src/tests/config.test.ts` covering boundary values for each bounded field (at min, at max, below min, above max). Update any existing tests in `src/tests/cli.test.ts` that reference `MAX_CONCURRENCY` to use the new bounds.

## Integration Points

- **`src/config.ts` — `DispatchConfig` interface:** Must add `planTimeout?: number` and `concurrency?: number` fields. The `CONFIG_KEYS` tuple must include the new keys. The `ConfigKey` type is derived from `CONFIG_KEYS` and will update automatically.
- **`src/config.ts` — `validateConfigValue()` function:** Must add new `case` branches for `"planTimeout"` and `"concurrency"`, and update the `"testTimeout"` branch to use `CONFIG_BOUNDS`.
- **`src/cli.ts` — CLI argument parsing:** The `--concurrency` and `--plan-timeout` parsing blocks should reference `CONFIG_BOUNDS` for their range checks. `MAX_CONCURRENCY` (currently `64`) should be updated or re-derived from `CONFIG_BOUNDS.concurrency.max`. Since `src/tests/cli.test.ts` imports `MAX_CONCURRENCY`, maintain the export.
- **`src/orchestrator/runner.ts` — `RawCliArgs` interface:** Already has `concurrency?: number` and `planTimeout?: number` fields — no changes needed.
- **`src/orchestrator/cli-config.ts`:** Handles config merging — may need to merge `planTimeout` and `concurrency` from config file into resolved args, following the same pattern as `provider`, `model`, `source`.
- **`tsup.config.ts`:** Single-line target change from `"node18"` to `"node20"`.
- **`tsconfig.json`:** Verify ES2022 target is compatible with Node 20 (it is — no change expected).
- **Vitest test conventions:** Tests use `describe`/`it`/`expect` from Vitest. Config tests use temp directories with `mkdtemp` and clean up with `afterEach`. Follow the existing pattern in `src/tests/config.test.ts`.
- **Build verification:** `npm run build` (tsup) must pass. `npm run test` (vitest run) must pass.

## Tasks

- [x] (P) Add `CONFIG_BOUNDS` constant and extend `DispatchConfig` interface, `CONFIG_KEYS`, and `validateConfigValue()` in `src/config.ts` to support `planTimeout` and `concurrency` with upper-bound range validation for all three numeric fields (`testTimeout`, `planTimeout`, `concurrency`). Export `CONFIG_BOUNDS` so it can be consumed by CLI parsing. Commit with message: `feat(config): add upper-bound validation for testTimeout, planTimeout, and concurrency`.
- [x] (P) Change `target: "node18"` to `target: "node20"` in `tsup.config.ts` to align the build target with the minimum Node.js runtime declared in `package.json` (`>=20.12.0`). Commit with message: `fix(build): align tsup target with minimum Node.js 20 runtime`.
- [x] (S) Refactor `src/cli.ts` to import and use `CONFIG_BOUNDS` from `src/config.ts` for `--concurrency` and `--plan-timeout` range validation, replacing the local `MAX_CONCURRENCY` constant. Keep `MAX_CONCURRENCY` exported (re-derived from `CONFIG_BOUNDS.concurrency.max`) for backward compatibility with existing test imports. Update any CLI config merging logic in `src/orchestrator/cli-config.ts` if needed to merge `planTimeout` and `concurrency` from the config file. Commit with message: `refactor(cli): use shared CONFIG_BOUNDS for numeric argument validation`.
- [x] (S) Add unit tests in `src/tests/config.test.ts` for boundary-value validation of `testTimeout`, `planTimeout`, and `concurrency` — covering values at min (1), at max (120 or 32), below min (0, -1), above max (121 or 33, 999), and non-numeric inputs. Update any affected assertions in `src/tests/cli.test.ts` that reference `MAX_CONCURRENCY` or the old validation behavior. Commit with message: `test(config): add boundary-value tests for numeric config validation`.
- [x] (I) Run the full build (`npm run build`) and test suite (`npm run test`) to verify all changes pass without new warnings or errors.

## References

- Node.js 20 ES compatibility: https://node.green/
- tsup `target` option: https://tsup.egoist.dev/#target-environment
- Vitest documentation: https://vitest.dev/

### PR DESCRIPTION
**Pull request by PatrickRuddiman**
Closes #170

## Summary

Adds upper-bound range validation to all numeric configuration fields (`testTimeout`, `planTimeout`, `concurrency`) and fixes the tsup build target to match the minimum Node.js runtime declared in `package.json`.

## Changes

**Config validation (`src/config.ts`)**
- Add `CONFIG_BOUNDS` constant defining min/max ranges for `testTimeout` (1–120), `planTimeout` (1–120), and `concurrency` (1–32)
- Extend `DispatchConfig` interface and `CONFIG_KEYS` to include `planTimeout` and `concurrency`
- Update `validateConfigValue()` to enforce range bounds for all three numeric fields (previously only checked for positive values on `testTimeout`)

**CLI argument parsing (`src/cli.ts`)**
- Import and use shared `CONFIG_BOUNDS` for `--concurrency` and `--plan-timeout` validation, replacing hardcoded limits
- Re-derive `MAX_CONCURRENCY` from `CONFIG_BOUNDS.concurrency.max` for backward compatibility

**Config merging (`src/orchestrator/cli-config.ts`)**
- Add `planTimeout` and `concurrency` to the config-to-CLI key mapping so config file values are properly merged

**Build target (`tsup.config.ts`)**
- Change `target: "node18"` → `target: "node20"` to match the `engines: { node: ">=20.12.0" }` constraint in `package.json`

**Tests**
- Add comprehensive boundary-value tests in `src/tests/config.test.ts` for all three numeric fields (min, max, below-min, above-max, non-numeric, Infinity, NaN)
- Update `src/tests/cli.test.ts` to mock `CONFIG_BOUNDS` and use dynamic max values in assertions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,308 +1,266 @@
-# dispatch
+# Dispatch
 
-AI agent orchestration CLI — parse markdown task files, dispatch each unit of work to code agents (OpenCode, GitHub Copilot, etc.), and commit results with conventional commits.
+AI agent orchestration CLI — parse work items from GitHub Issues, Azure DevOps, or local markdown files, dispatch each unit of work to a coding agent (OpenCode, GitHub Copilot, Claude Code, or OpenAI Codex), and commit results with conventional commits.
```
[the remark this answers is not in the document]
Comment (PatrickRuddiman) [reply README.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/PatrickRuddiman/Dispatch/pull/198#discussion_r2880544327)
